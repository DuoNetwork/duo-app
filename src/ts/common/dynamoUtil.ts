import { DynamoUtil } from '@finbook/duo-market-data';

let config = {};
try {
	config = require(`../keys/aws.ui.${__KOVAN__ ? 'dev' : 'live'}.json`);
} catch (error) {
	console.log(error);
}

const dynamoUtil = new DynamoUtil(config, !__KOVAN__, '');

export default dynamoUtil;
