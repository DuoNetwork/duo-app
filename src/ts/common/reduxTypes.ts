import { IAssets, IPriceData, ITimeSeriesData } from './types';
export interface IState {
	ui: IUIState;
}

export interface IUIState {
	eth: ITimeSeriesData[];
	classA: ITimeSeriesData[];
	classB: ITimeSeriesData[];
	reset: ITimeSeriesData[];
	history: string[];
	message: {
		type: string;
		content: string;
		visible: boolean;
	};
	setting: {
		couponRate: number;
		upwardResetLimit: number;
		downwardResetLimit: number;
		periodicResetLimit: number;
	};
	form: {
		type: string;
		visible: boolean;
	};
	mv: ITimeSeriesData[];
	assets: IAssets;
	resetPrice: number;
	beta: number;
	day: number;
	upward: number;
	downward: number;
	periodic: number;
	price: IPriceData;
}

export type Action =
	| IBaseAction
	| IBooleanAction
	| IStringAction
	| IObjectAction
	| INumberAction
	| ITradeAction;

export type Dispatch = (action: Action | ThunkAction | PromiseAction) => any;

export interface IBaseAction {
	type: string;
}

export interface IBooleanAction {
	type: string;
	value: boolean;
}

export interface IStringAction {
	type: string;
	value: string;
}

export interface IObjectAction {
	type: string;
	value: object;
}

export interface INumberAction {
	type: string;
	value: number;
}

export interface ITradeAction {
	type: string;
	history: string;
	assets: IAssets;
}

export type ThunkAction = (dispatch: Dispatch) => any;
export type PromiseAction = Promise<Action>;
