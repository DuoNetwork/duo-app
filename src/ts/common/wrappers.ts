import infura from 'ts/keys/infura.json';
import BeethovenWapper from '../../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
export const web3Wrapper = new Web3Wrapper(window, '', provider, !__KOVAN__);
export const beethovenWapper = new BeethovenWapper(web3Wrapper);
