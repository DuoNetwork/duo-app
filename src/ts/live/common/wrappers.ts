import BeethovenWapper from '../../../../../duo-contract-wrapper/src/BeethovenWapper';
import Web3Wrapper from '../../../../../duo-contract-wrapper/src/Web3Wrapper';
import infura from '../../keys/infura.json';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
export const web3Wrapper = new Web3Wrapper(window, '', provider, !__KOVAN__);
export const beethovenWapper = new BeethovenWapper(web3Wrapper);
