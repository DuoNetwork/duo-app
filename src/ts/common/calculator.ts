import moment from 'moment';
import * as CST from './constants';
import { IUIState } from './reduxTypes';
import { IAssets, IRawData, ITimeSeriesData } from './types';

class Calculator {
	public getAllTimeSeriesFromEth(
		rawData: IRawData[],
		alpha: number = 1,
		coupon: number = 0.0002,
		upperLimit: number = 2,
		lowerLimit: number = 0.25,
		periodLimit: number = 1.02
	) {
		const eth: ITimeSeriesData[] = [],
			classA: ITimeSeriesData[] = [],
			classB: ITimeSeriesData[] = [],
			resetPrice: ITimeSeriesData[] = [],
			beta: ITimeSeriesData[] = [],
			upward: ITimeSeriesData[] = [],
			downward: ITimeSeriesData[] = [],
			periodic: ITimeSeriesData[] = [];
		let count = -1;
		let lastResetDatetime = moment(rawData[0].Date, 'YYYY-MM-DD').valueOf();
		let lastResetPrice = rawData[0].Price;
		let currentBeta = 1;
		rawData.forEach(d => {
			count++;
			const datetime = moment(d.Date, 'YYYY-MM-DD').valueOf();
			const ethPx = d.Price;
			eth.push({
				datetime: datetime,
				value: ethPx
			});

			let [navA, navB] = this.calculateNav(
				ethPx,
				lastResetPrice,
				alpha,
				currentBeta,
				count,
				coupon
			);

			if (navB >= upperLimit || navB <= lowerLimit || navA >= periodLimit) {
				lastResetPrice = ethPx;
				lastResetDatetime = datetime;
				count = 0;
				if (navB >= upperLimit)
					upward.push({
						datetime: datetime,
						value: 1
					});
				else if (navB <= lowerLimit)
					downward.push({
						datetime: datetime,
						value: 1
					});
				else
					periodic.push({
						datetime: datetime,
						value: 1
					});

				// class B does not reset in periodic reset
				if (navA < periodLimit) {
					classB.push({
						datetime: datetime,
						value: navB
					});
					navB = 1;
					currentBeta = 1;
				} else
					currentBeta = this.updateBeta(currentBeta, ethPx, lastResetPrice, navA, alpha);

				classA.push({
					datetime: datetime,
					value: navA
				});

				navA = 1;
			}

			classA.push({
				datetime: datetime,
				value: navA
			});

			classB.push({
				datetime: datetime,
				value: navB
			});

			resetPrice.push({
				datetime: lastResetDatetime,
				value: lastResetPrice
			});

			beta.push({
				datetime: datetime,
				value: currentBeta
			});
		});
		return [eth, classA, classB, resetPrice, beta, upward, downward, periodic];
	}

	public calculateNav(
		price: number,
		resetPrice: number,
		alpha: number,
		beta: number,
		periodCount: number,
		coupon: number
	) {
		let navParent = price / resetPrice / beta * (1 + alpha);

		let navA = 1 + periodCount * coupon;
		let navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public updateBeta(
		prevBeta: number,
		price: number,
		resetPrice: number,
		navA: number,
		alpha: number
	) {
		return (
			(1 + alpha) * price / ((1 + alpha) * price - resetPrice * alpha * prevBeta * (navA - 1))
		);
	}

	public assetUpwardReset(assets: IAssets, eth: number, navA: number, navB: number): IAssets {
		return {
			ETH: assets.ETH + ((navA - 1) * assets.ClassA + (navB - 1) * assets.ClassB) / eth,
			ClassA: assets.ClassA,
			ClassB: assets.ClassB
		};
	}

	public assetDownwardReset(assets: IAssets, eth: number, navA: number, navB: number): IAssets {
		return {
			ETH: assets.ETH + (navA - navB) * assets.ClassA / eth,
			ClassA: assets.ClassA * navB,
			ClassB: assets.ClassB * navB
		};
	}

	public assetPeriodicReset(assets: IAssets, eth: number, navA: number): IAssets {
		return {
			ETH: assets.ETH + (navA - 1) * assets.ClassA / eth,
			ClassA: assets.ClassA,
			ClassB: assets.ClassB
		};
	}

	public calculateNextDayState(state: IUIState): object {
		const { eth, classA, classB, assets, day, upward, downward, periodic, mv } = state;
		const newDayCount = day + 1;
		if (newDayCount >= eth.length) return {};
		else {
			const newDatetime = eth[newDayCount].datetime;
			const newEthPx = eth[newDayCount].value;
			const upwardSoFar = upward.filter(d => d.datetime <= newDatetime);
			const downwardSoFar = downward.filter(d => d.datetime <= newDatetime);
			const periodicSoFar = periodic.filter(d => d.datetime <= newDatetime);
			const isUpward =
				upwardSoFar.length && upwardSoFar[upwardSoFar.length - 1].datetime === newDatetime;
			const isDownward =
				downwardSoFar.length &&
				downwardSoFar[downwardSoFar.length - 1].datetime === newDatetime;
			const isPeriodic =
				periodicSoFar.length &&
				periodicSoFar[periodicSoFar.length - 1].datetime === newDatetime;
			const upwardCount = upwardSoFar.length - (isUpward ? 1 : 0);
			const downwardCount = downwardSoFar.length - (isDownward ? 1 : 0);
			const periodicCount = periodicSoFar.length - (isPeriodic ? 1 : 0);
			const newNavA = classA[newDayCount + upwardCount + downwardCount + periodicCount].value;
			const newNavB = classB[newDayCount + upwardCount + downwardCount].value;

			let newAssets: IAssets;
			let msg = '';
			if (isUpward) {
				newAssets = calculator.assetUpwardReset(assets, newEthPx, newNavA, newNavB);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (upward) triggered.</div>";
			} else if (isDownward) {
				newAssets = calculator.assetDownwardReset(assets, newEthPx, newNavA, newNavB);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (downward) triggered.</div>";
			} else if (isPeriodic) {
				newAssets = calculator.assetPeriodicReset(assets, newEthPx, newNavA);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (periodic) triggered.</div>";
			} else newAssets = assets;

			return Object.assign(
				{
					[CST.AC_MV]: [
						...mv,
						{
							datetime: newDatetime,
							value:
								assets.ETH * newEthPx +
								assets.ClassA * newNavA +
								assets.ClassB * newNavB
						}
					],
					[CST.AC_ASSETS]: newAssets,
					[CST.AC_DAY]: newDayCount,
					[CST.AC_PRICE]: {
						Date: newDatetime,
						ETH: newEthPx,
						ClassA:
							classA[
								newDayCount +
									upwardSoFar.length +
									downwardSoFar.length +
									periodicSoFar.length
							].value,
						ClassB:
							classB[newDayCount + upwardSoFar.length + downwardSoFar.length].value
					}
				},
				msg
					? {
							[CST.AC_MESSAGE]: {
								type: "<div style='color: rgba(0,186,255,0.7)'>INFORMATION</div>",
								content: msg,
								visible: true
							}
					}
					: {}
			);
		}
	}

	public calculateForwardState(state: IUIState): object {
		const { eth, classA, classB, assets, day, upward, downward, periodic, mv } = state;
		const currentDatetime = eth[day].datetime;
		const endDatetime = eth[eth.length - 1].datetime;
		const upwardSoFar = upward.filter(d => d.datetime <= currentDatetime);
		const downwardSoFar = downward.filter(d => d.datetime <= currentDatetime);
		const periodicSoFar = periodic.filter(d => d.datetime <= currentDatetime);
		const nextUpward = upward[upwardSoFar.length]
			? upward[upwardSoFar.length].datetime
			: endDatetime;
		const nextDownward = downward[downwardSoFar.length]
			? downward[downwardSoFar.length].datetime
			: endDatetime;
		const nextPeriodic = periodic[periodicSoFar.length]
			? periodic[periodicSoFar.length].datetime
			: endDatetime;
		const nextResetDate = Math.min(nextUpward, nextDownward, nextPeriodic);
		let curDate = day + 1;
		let newMV = [...mv];
		let newDatetime = eth[curDate].datetime;
		if (newDatetime === nextResetDate) return this.calculateNextDayState(state);

		let newEthPx, newNavA, newNavB;
		while (curDate < eth.length && eth[curDate].datetime < nextResetDate) {
			newDatetime = eth[curDate].datetime;
			newEthPx = eth[curDate].value;
			newNavA =
				classA[curDate + upwardSoFar.length + downwardSoFar.length + periodicSoFar.length]
					.value;
			newNavB = classB[curDate + upwardSoFar.length + downwardSoFar.length].value;
			newMV.push({
				datetime: newDatetime,
				value: assets.ETH * newEthPx + assets.ClassA * newNavA + assets.ClassB * newNavB
			});
			curDate++;
		}
		curDate--;
		return {
			[CST.AC_MV]: newMV,
			[CST.AC_DAY]: curDate,
			[CST.AC_PRICE]: {
				Date: newDatetime,
				ETH: newEthPx,
				ClassA:
					classA[
						curDate + upwardSoFar.length + downwardSoFar.length + periodicSoFar.length
					].value,
				ClassB: classB[curDate + upwardSoFar.length + downwardSoFar.length].value
			}
		};
	}
}

const calculator = new Calculator();
export default calculator;
