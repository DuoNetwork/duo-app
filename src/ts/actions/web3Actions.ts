import * as CST from 'ts/common/constants';
import { VoidThunkAction } from 'ts/common/types';
import { web3Wrapper } from 'ts/common/wrappers';

export function accountUpdate(account: string) {
	return {
		type: CST.AC_WEB3_ACCOUNT,
		value: account
	};
}

export function getAccount(): VoidThunkAction {
	return async dispatch => dispatch(accountUpdate(await web3Wrapper.getCurrentAddress()));
}

export function networkUpdate(networkId: number) {
	return {
		type: CST.AC_WEB3_NETWORK,
		value: networkId
	};
}

export function getNetwork(): VoidThunkAction {
	return async dispatch => dispatch(networkUpdate(await web3Wrapper.getCurrentNetwork()));
}

export function gasPriceUpdate(gasPrice: number) {
	return {
		type: CST.AC_WEB3_GAS_PX,
		value: gasPrice
	};
}

export function getGasPrice(): VoidThunkAction {
	return async dispatch => dispatch(gasPriceUpdate(await web3Wrapper.getGasPrice()));
}

export function balanceUpdate(balance: number) {
	return {
		type: CST.AC_WEB3_BALACE,
		value: balance
	};
}

export function getBalance(): VoidThunkAction {
	return async (dispatch, getState) =>
		dispatch(balanceUpdate(await web3Wrapper.getEthBalance(getState().web3.account)));
}

export function refresh(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getAccount());
		dispatch(getGasPrice());
		dispatch(getNetwork());
		dispatch(getBalance());
	};
}
