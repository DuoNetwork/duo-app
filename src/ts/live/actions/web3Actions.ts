import * as CST from '../common/constants';
import contract from '../common/contract';
import { VoidThunkAction } from '../common/types';

export function accountUpdate(account: string) {
	return {
		type: CST.AC_ACCOUNT,
		value: account
	};
}

export function getAccount(): VoidThunkAction {
	return async dispatch => dispatch(accountUpdate(await contract.getCurrentAddress()));
}

export function networkUpdate(networkId: number) {
	return {
		type: CST.AC_NETWORK,
		value: networkId
	};
}

export function getNetwork(): VoidThunkAction {
	return async dispatch => dispatch(networkUpdate(await contract.getCurrentNetwork()));
}

export function gasPriceUpdate(gasPrice: number) {
	return {
		type: CST.AC_GAS_PX,
		value: gasPrice
	};
}

export function getGasPrice(): VoidThunkAction {
	return async dispatch => dispatch(gasPriceUpdate(await contract.getGasPrice()));
}

export function refresh(): VoidThunkAction {
	return dispatch => {
		dispatch(getAccount());
		dispatch(getGasPrice());
		dispatch(getNetwork());
	};
}
