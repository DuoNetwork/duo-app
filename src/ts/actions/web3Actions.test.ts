import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { web3Wrapper } from 'ts/common/wrappers';
import * as web3Actions from './web3Actions';

const mockStore = configureMockStore([thunk]);

describe('actions', () => {
	test('accountUpdate', () => {
		expect(web3Actions.accountUpdate('test')).toMatchSnapshot();
	});

	test('getAccount', () => {
		const store: any = mockStore({});
		web3Wrapper.getCurrentAddress = jest.fn(() => Promise.resolve('0x0'));
		store.dispatch(web3Actions.getAccount());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('networkUpdate', () => {
		expect(web3Actions.networkUpdate(123)).toMatchSnapshot();
	});

	test('getNetwork', () => {
		const store: any = mockStore({});
		web3Wrapper.getCurrentNetwork = jest.fn(() => Promise.resolve(123));
		store.dispatch(web3Actions.getNetwork());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('gasPriceUpdate', () => {
		expect(web3Actions.gasPriceUpdate(123)).toMatchSnapshot();
	});

	test('getGasPrice', () => {
		const store: any = mockStore({});
		web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(456));
		store.dispatch(web3Actions.getGasPrice());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('balanceUpdate', () => {
		expect(web3Actions.balanceUpdate(123)).toMatchSnapshot();
	});

	test('getBalance', () => {
		const store: any = mockStore({
			web3: {
				account: '0x0'
			}
		});
		web3Wrapper.getEthBalance = jest.fn(() => Promise.resolve(456));
		store.dispatch(web3Actions.getBalance());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
