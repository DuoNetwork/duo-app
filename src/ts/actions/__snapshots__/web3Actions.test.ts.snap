// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions accountUpdate 1`] = `
Object {
  "type": "web3Account",
  "value": "test",
}
`;

exports[`actions balanceUpdate 1`] = `
Object {
  "type": "web3Balance",
  "value": 123,
}
`;

exports[`actions gasPriceUpdate 1`] = `
Object {
  "type": "web3GasPrice",
  "value": 123,
}
`;

exports[`actions getAccount 1`] = `
Array [
  Object {
    "type": "web3Account",
    "value": "0x0",
  },
]
`;

exports[`actions getBalance 1`] = `
Array [
  Object {
    "type": "web3Balance",
    "value": 456,
  },
]
`;

exports[`actions getGasPrice 1`] = `
Array [
  Object {
    "type": "web3GasPrice",
    "value": 456,
  },
]
`;

exports[`actions getNetwork 1`] = `
Array [
  Object {
    "type": "web3Network",
    "value": 123,
  },
]
`;

exports[`actions networkUpdate 1`] = `
Object {
  "type": "web3Network",
  "value": 123,
}
`;
